### PR TITLE
datadriven: add rewrite mode

### PR DESCRIPTION
**Pull request by accelsao**
1. add rewrite
2. don't trim on expected value (bug)
3. move the test to independent file

### CODE REVIEW

**Comment by BusyJay** [comment]
Is rewrite mode used in the raft tests?

**Comment by accelsao** [comment]
This is like a tool, not really necessary for testing.
If we change the `Display` in the future, we can just modify the test by hand.
I forget we need group commit so I skip this feature at the beginning.

File: datadriven/src/testdata/multiline [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+large
+----
+----
+more
+than
+five
+lines
+of
+
```
Comment (BusyJay) [inline datadriven/src/testdata/multiline:17]: There is no blank line in the original test data.

File: datadriven/src/test_data_reader.rs [hunk 2]
```diff
@@ -10,10 +10,12 @@ pub struct TestDataReader<'a> {
     pub data: TestData,
     scanner: Enumerate<Lines<'a>>,
     pub logger: slog::Logger,
+    pub rewrite: bool,
+    pub rewrite_buffer: String,
```
Comment (BusyJay) [inline datadriven/src/test_data_reader.rs:14]: `Option<String>` seems a more safe definition. People can forget to check the Boolean flag.

File: datadriven/src/test_data_reader.rs [hunk 3]
```diff
@@ -34,6 +38,7 @@ impl<'a> TestDataReader<'a> {
 
             let mut pos = line.unwrap().0;
             let mut line = line.unwrap().1.to_string();
+            self.emit(&line);
```
Comment (BusyJay) [inline datadriven/src/test_data_reader.rs:41]: I think you need to trim after emitting. For example, both of following two lines are comments and should be ignored.
```
# comment
   # comment
```

File: datadriven/src/test_data_reader.rs [hunk 4]
```diff
@@ -134,42 +139,66 @@ impl<'a> TestDataReader<'a> {
...
                         }
-                        line2 += "\n";
-                        self.data.expected.push_str(&line2);
+
+                        break;
                     }
-                    line += "\n";
-                    self.data.expected.push_str(&line);
+                    line2 += "\n";
+                    self.data.expected.push_str(&line2);
```
Comment (BusyJay) [inline datadriven/src/test_data_reader.rs:175]: This seems wrong. Original implement push `line` first and then `line2`.

File: datadriven/src/test_data.rs [hunk 5]
```diff
@@ -22,7 +22,21 @@ impl fmt::Display for CmdArg {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
         match self.vals.len() {
             0 => write!(f, "{}", self.key),
-            _ => write!(f, "{}={:?}", self.key, self.vals),
+            1 => {
+                let mut str = format!("{}=", self.key);
```
Comment (BusyJay) [inline datadriven/src/test_data.rs:26]: Why not just `write!(f, "{}={}", self.key, self.vals[0])`?

File: datadriven/src/test_data.rs [hunk 6]
```diff
@@ -22,7 +22,21 @@ impl fmt::Display for CmdArg {
...
+                str += &self.vals[0];
+                write!(f, "{}", str)
+            }
+            _ => {
+                let mut str = format!("{}=(", self.key);
```
Comment (BusyJay) [inline datadriven/src/test_data.rs:31]: Why not `write!(f, "{}=({})", self.key, self.vals.join(", "))`?

File: datadriven/src/testdata/multiline (hunk 1 above)
Comment (BusyJay) [reply datadriven/src/testdata/multiline:17]: Any updates?

File: datadriven/src/datadriven.rs [hunk 7]
```diff
@@ -71,233 +78,102 @@ use std::path::{Path, PathBuf};
...
+    let mut r = TestDataReader::new(source_name, content, rewrite, logger);
+
     while r.next()? {
-        run_directive(&r, f)?;
+        run_directive(&mut r, f)?;
+    }
+
+    if r.rewrite {
+        // remove redundant '\n'
+        let data = r.rewrite_buffer.map(|rb| rb.trim_end().to_string() + "\n");
```
Comment (BusyJay) [inline datadriven/src/datadriven.rs:126]: How about:
```
r.rewrite_buffer.map(|mut rb| {
    if rb.ends_with("\n\n") {
        rb.pop();
    }
    rb
}
```

File: datadriven/src/test_data.rs (hunk 6 above)
Comment (accelsao) [reply datadriven/src/test_data.rs]: Cool!

**Comment by BusyJay** [comment]
@NingLin-P @Fullstop000 PTAL

**Comment by NingLin-P** [comment]
LGTM

File: datadriven/src/datadriven.rs [hunk 8]
```diff
@@ -71,233 +78,105 @@ use std::path::{Path, PathBuf};
...
-        run_directive(&r, f)?;
+        run_directive(&mut r, f)?;
     }
-    Ok(())
+
+    // remove redundant '\n'
+    let data = r.rewrite_buffer.map(|mut rb| {
+        if rb.ends_with("\n\n") {
+            rb.pop();
+            rb.pop();
```
Comment (BusyJay) [inline datadriven/src/datadriven.rs:128]: Why? I can see go implement will still keep one newline.

Comment (accelsao) [reply datadriven/src/datadriven.rs:128]: There is no "\n" at the end of rewrite data, so I need to pop 2 "\n"

Comment (accelsao) [reply datadriven/src/datadriven.rs]: the addition "\n" is the reason why I pop 2 "\n"

Comment (accelsao) [reply datadriven/src/datadriven.rs]: I am not sure why they dont have any blank line at the end

**Pull request merged**